### PR TITLE
How to setup, containerize and test a single Hadoop cluster using Docker and docker-compose

### LINKED ISSUE #2543: How to set up a single Hadoop cluster using Docker
We ask that you please be patient as our team works through approving and publishing all articles/tutorials in a timely manner. Allow 1-3 days for a topic to be reviewed and/or approved - allow 3-7 days for an articles to be reviewed and/or published.

Be sure to visit our [Resources Page](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md) for tools, resources, and example articles to go over.

**NOTE: (Please ensure that you have only one open issue + linked pull request at a time. This will ensure that we complete the article in a timely manner from inception to publishing.)**

**NOTE: We tend to stray away or tend not to publish reviews/comparisons of commercial product offerings.**

*We're looking for the first 2-3 paragraphs of the article that appropriately summarize what your article will be about.*

### Proposed title of article
Your title should be descriptive of the article/tutorial. Be Specific. *Use keyword research to gain your article higher ranking.*

How to set up a single Hadoop cluster using Docker

### Introduction paragraph (2-3 paragraphs):

This guide will teach you how to set up a single Hadoop cluster using Docker. But before that, let's take a recap of what the Docker and Hadoop cluster is.

### What is a Hadoop cluster?

The Hadoop framework helps in processing and analyzing Big Data. Hadoop framework stores and processes vast amounts of any data efficiently using a cluster of computer hardware. Thus we can refer Hadoop cluster as a special type of computer cluster. It is designed for storing and analyzing vast amounts of unstructured or structured data. Clustering is mainly used for parallel processing, load balancing, and fault tolerance. a Hadoop cluster run Hadoop open-source distributed processing software.

A Hadoop cluster works the same as a computer cluster. A computer cluster can be defined as a set of connected computers that work together in a way that they could be viewed as a single system. This way, it is relatively easy to add more power to the cluster. You simply add a new computer to the network.

The Hadoop ecosystem comprises three main components that are designed to specifically work on big data this are;

- Hadoop distributed file system (HDFS)

HDFS is a storage unit that helps to distribute data amongst many computers and is stored in blocks.  Hadoop framework divides the data into smaller chunks and stores each part of the data on a separate node within the cluster. Let's say you have around 4 terabytes of data. HDFS will split the data into multiple data blocks and then store the data in several data nodes in the Hadoop cluster. With each block(node) assigned 128 megabytes as the default size. This way time taken to store this data onto the disk is significantly reduced. The total time used to store this data into a disk will be equal to the total time used to store data in one block. HDFS will then use the replication method and make copies of each data node to other machines present within the cluster. This ensures no data is lost even if one data node crashes, making HDFS fault-tolerant.

- MapReduce

MapReduce splits data into parts and processes each of them separately on different data nodes. The individual results are then aggregated to give the final output.

- Yet Another Resource Negotiator (YARN)

YARN manages resources within a cluster. It consists of a resource manager, node manager, application master, and containers. The resource manager assigns resources to the Hadoop cluster. Node managers handle the nodes and monitor the resource usage in the node. The containers hold a collection of physical resources.  With YARN, the application requests the container from the node manager. Once the node manager gets the resources, it sends them to the resource manager. This way, YARN processes job requests and manages cluster resources in a Hadoop cluster.

### Key takeaways: 
What is a Hadoop cluster?
What is Docker and Docker containers?
How to set up a Hadoop cluster using docker-compose
Testing the Hadoop cluster

### References: 
Please list links to any published content/research that you intend to use to support/guide this article.

### Templates to use as guides
- [How To Guide Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/how-to-guide-template.md)
- [Software Review Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/software-review-template.md)
- [Tutorial Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/tutorial-template.md)

### PR DESCRIPTION
**Pull request by Antony-Prince**
…-docker-compose

### Article checklist

**ATTENTION: In your PR - add the ISSUE (topic form) # that this PR corresponds to. You can do this by also adding a comment that states **["This fixes #(enter your Idea Content Form #)"](https://youtu.be/TKJ4RdhyB5Y?t=278)** to link and close your Idea Content (issue) when the PR is merged.**
      
      
**NOTE: (Please ensure that you have only one open issue + linked pull request at a time. This will ensure that we complete the article in a timely manner from inception to publishing.)**

**If you have not already please go over our [Resources Page](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md) for more tips and tricks on how to improve our overall technical writing.**

#### Preliminary Checks - Formatting and Structure 
- [x] Does your article follow any of the suggested structure formats? - **see [example formats](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md#example-format-structure)**
- [x] Is your article properly formatted in Markdown? - **see [Github Markdown guide](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet)**
- [x] Have you used the correct folder and file structure? - **see [Contribution Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**
- [x] Is your article over 750 words? - **Tool**: [Wordcounter.net](https://wordcounter.net)
- [x] Have you used our preferred heading structure? - **[H3 (###) and up](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#headers)**
- [x] Does your article provide enough value and detail about your topic? - **Articles should be clear, accurate, and fully explained.**
- [x] Can your article be understood by beginners? - **Assume the audience is smart but has no prior exposure to the common terminology in your article.**
- [x] Have you included a hero image that is 600x400 pixels and have the copyright to use it? - **Tools:** [Pixlr Image Editor](https://pixlr.com/e) for resizing and [Unsplash](https://unsplash.com) for Creative Commons images
- [x] In your PR - add a comment that states **["This fixes #(enter your Idea Content Form #)"](https://youtu.be/TKJ4RdhyB5Y?t=278)** to link and close your Idea Content (issue) when the PR is merged.
      
#### Grammar & plagiarism checks
- [x] Have you spellchecked and grammar checked your article? - **Tools:** [Write&Improve](https://writeandimprove.com/?lang=en-GB) - [ProWritingAid](https://prowritingaid.com/) -  [Grammarly](https://grammarly.com) 
- [x] Please place your article through a 3rd party plagiarism checker? We suggest using Quetext, [this tool](https://www.quetext.com/) is free and has a daily limit. **We typically accept articles with 10% or less.**
- [x] Have you checked your article for readability? - **Tool:** [Hemmingway](http://www.hemingwayapp.com)
- [x] Have you added sources for quotes and images that aren't yours?

#### Technical checks
- [x] Are your code snippets properly formatted for syntax highlighting - **see [Syntax guide](https://gohugo.io/content-management/syntax-highlighting)**

### Contribution guidelines
For first-time contributors and for more details, **see [Contributing Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**
This fixes #2543

### CODE REVIEW

**Comment by Bayler** [comment]
Hi @Antony-Prince Welcome to Section engineering program. We are pleased have you and congrats on submitting your first article successfully. I will be reviewing your article. 

Before I continue with the review, you need to do a couple of thing first:
1. On the article checklist, it has indicated that you need to upload a hero image. Please do that. Make sure to save it as hero.jpg/hero.png etc. Also make sure it is 600px by 400px in dimension.
2. We need to create an author profile for you. Thus, you need to add a brief Author Bio (index.md file) and Headshot image (avatar file i.e., avatar.jpg). Include these files in the authors folder which is inside the 'contents' folder. This will help so that you can be credited for your hard work.

After you've finished that, I can proceed with the review.

**Comment by Antony-Prince** [comment]
> Hi @Antony-Prince Welcome to Section engineering program. We are pleased have you and congrats on submitting your first article successfully. I will be reviewing your article.
> 
> Before I continue with the review, you need to do a couple of things first:
> 
> 1. On the article checklist, it has been indicated that you need to upload a hero image. Please do that. Make sure to save it as hero.jpg/hero.png etc. Also make sure it is 600px by 400px in dimension.
> 2. We need to create an author profile for you. Thus, you need to add a brief Author Bio (index.md file) and Headshot image (avatar file i.e., avatar.jpg). Include these files in the authors folder which is inside the 'contents' folder. This will help so that you can be credited for your hard work.
> 
> After you've finished that, I can proceed with the review.

Hello @Bayler This is my second article, I had one published already

**Comment by Bayler** [comment]
Oh. Thought it was your first. Sorry. You can add a hero image then

**Comment by Antony-Prince** [comment]
> Oh. Thought it was your first. Sorry. You can add a hero image then

Added @Bayler

**Comment by Bayler** [comment]
@Antony-Prince 

"How to setup, containerize and test a single Hadoop cluster using Docker and docker-compose" is a much cleaner topic to use than the current one.

You can add the "What is Hadoop?" as a header inside your article instead. First start describing what Hadoop is then start on Hadoop clusters

cc: @hectorkambow

**Comment by Antony-Prince** [comment]
> What is Hadoop?

Yeah, that reflects well. I have updated that @Bayler

File: content/articles/set-up-containerize-and-test-a-single-hadoop-cluster-using-docker-and-docker-compose/index.md [hunk 1]
```diff
@@ -0,0 +1,356 @@
...
+To begin, run the following command to get a Hadoop Docker image from the Docker hub libraries;
+
+```bash
+sudo docker pull sequenceiq/hadoop-docker:2.7.1
+```
+
+This will download Hadoop and set it inside the installed Docker engine on your computer. You can check whether the Hadoop Docker image was downloaded successfully by running;
+
+This will download the Hadoop image with its YARN properties such as the node manager, resource manager, and history server and install it in your computer's Docker engine. Run the below command to see if the Hadoop Docker image was successfully downloaded.
+
```
Comment (Bayler) [inline content/articles/set-up-containerize-and-test-a-single-hadoop-cluster-using-docker-and-docker-compose/index.md:274]: I believe there is a piece of code that you have forgotten to include here.

File: content/articles/set-up-containerize-and-test-a-single-hadoop-cluster-using-docker-and-docker-compose/index.md [hunk 2]
```diff
@@ -0,0 +1,356 @@
...
+```bash
+docker ps
+```
+
+![hadoop-running-containers-response](/engineering-education/set-up-containerize-and-test-a-single-hadoop-cluster-using-docker-and-docker-compose/hadoop-running-containers-response.png)
+
+### Check the running Hadoop containerized environment
+
+To visualize the running Hadoop application, you need to first the containers running IP address and test it on the browser using the mapped container port. Run this command to get your IP address.
+
```
Comment (Bayler) [inline content/articles/set-up-containerize-and-test-a-single-hadoop-cluster-using-docker-and-docker-compose/index.md:158]: This sentence is not clear. Please correct it.

**Comment by Bayler** [comment]
Hi @Antony-Prince This was a nice read. I like the way you've explained concepts and provided outputs with supported image examples.

However, I need you to work on some few corrections before we can proceed for final review:

1. Please add a 'Table of contents' and 'Prerequisite' section. It is good practice to include this inside your article. Use this [article](https://www.section.io/engineering-education/clip-connecting-texts-and-images/) as reference.
2. Once you're done with the corrections, please run your article on the [Hemmingway App](https://hemingwayapp.com/) to ensure the readability of your article is up to par. Currently, it's way higher. (Aim for a 9)  or below.

After you're done, we can proceed for final review.

**Comment by Antony-Prince** [comment]
> Hi @Antony-Prince This was a nice read. I like the way you've explained concepts and provided outputs with supported image examples.
> 
> However, I need you to work on some few corrections before we can proceed for final review:
> 
> 1. Please add a 'Table of contents' and 'Prerequisite' section. It is good practice to include this inside your article. Use this [article](https://www.section.io/engineering-education/clip-connecting-texts-and-images/) as reference.
> 2. Once you're done with the corrections, please run your article on the [Hemmingway App](https://hemingwayapp.com/) to ensure the readability of your article is up to par. Currently, it's way higher. (Aim for a 9)  or below.
> 
> After you're done, we can proceed for final review.

@Bayler I have added a 'Table of contents' and 'Prerequisite' section.  My Hemmingway score is 7 when I use the previewed index.md output.

**Comment by Bayler** [comment]
> > Hi @Antony-Prince This was a nice read. I like the way you've explained concepts and provided outputs with supported image examples.
> > However, I need you to work on some few corrections before we can proceed for final review:
> > 
> > 1. Please add a 'Table of contents' and 'Prerequisite' section. It is good practice to include this inside your article. Use this [article](https://www.section.io/engineering-education/clip-connecting-texts-and-images/) as reference.
> > 2. Once you're done with the corrections, please run your article on the [Hemmingway App](https://hemingwayapp.com/) to ensure the readability of your article is up to par. Currently, it's way higher. (Aim for a 9)  or below.
> > 
> > After you're done, we can proceed for final review.
> 
> @Bayler I have added a 'Table of contents' and 'Prerequisite' section. My Hemmingway score is 7 when I use the previewed index.md output.

Okay @Antony-Prince Let me check.

**Comment by Bayler** [comment]
> > Hi @Antony-Prince This was a nice read. I like the way you've explained concepts and provided outputs with supported image examples.
> > 
> > However, I need you to work on some few corrections before we can proceed for final review:
> > 
> > 1. Please add a 'Table of contents' and 'Prerequisite' section. It is good practice to include this inside your article. Use this [article](https://www.section.io/engineering-education/clip-connecting-texts-and-images/) as reference.
> > 2. Once you're done with the corrections, please run your article on the [Hemmingway App](https://hemingwayapp.com/) to ensure the readability of your article is up to par. Currently, it's way higher. (Aim for a 9)  or below.
> > 
> > After you're done, we can proceed for final review.
> 
> @Bayler I have added a 'Table of contents' and 'Prerequisite' section.  My Hemmingway score is 7 when I use the previewed index.md output.



The readability is still at 11.

**Comment by Antony-Prince** [comment]
@Bayler you are checking the readability  using the MD format
readability2

**Comment by Antony-Prince** [comment]
Am using the rendered/ previewed MD output.
readability

**Comment by Bayler** [comment]
> Am using the rendered/ previewed MD output.
> readability

@Antony-Prince I've noted where the discrepancy was coming in. Let me finalize the review.

**Comment by hectorkambow** [comment]
plagiarism check done

**Comment by ahmadmardeni1** [comment]
Great tutorial, good job 🎉@Antony-Prince @Bayler.

Please keep in mind that your article should always have an introduction.

**Pull request merged**